src/source/src/parser/mod.rs

```diff
@@ -88,51 +88,144 @@ pub struct SourceStreamChunkRowWriter<'a> {
+        builder.append_datum(&output).map_err(Into::into)
+    }
+
+    #[inline(always)]
+    fn rollback(builder: &mut ArrayBuilderImpl) {
+        builder.pop().unwrap()
+    }
+
+    #[inline(always)]
+    fn finish(writer: SourceStreamChunkRowWriter<'_>) {
+        writer.op_builder.push(Op::Insert)
+    }
+}
+
+struct OpActionDelete;
+
+impl OpAction for OpActionDelete {
+    type Output = Datum;
+
+    const DEFAULT_OUTPUT: Self::Output = None;
+
+    #[inline(always)]
+    fn apply(builder: &mut ArrayBuilderImpl, output: Datum) -> Result<()> {
+        builder.append_datum(&output).map_err(Into::into)
+    }
+
+    #[inline(always)]
+    fn rollback(builder: &mut ArrayBuilderImpl) {
+        builder.pop().unwrap()
+    }
+
+    #[inline(always)]
+    fn finish(writer: SourceStreamChunkRowWriter<'_>) {
+        writer.op_builder.push(Op::Delete)
+    }
+}
+
+struct OpActionUpdate;
+
+impl OpAction for OpActionUpdate {
+    type Output = (Datum, Datum);
+
+    const DEFAULT_OUTPUT: Self::Output = (None, None);
+
+    #[inline(always)]
+    fn apply(builder: &mut ArrayBuilderImpl, output: (Datum, Datum)) -> Result<()> {
+        builder.append_datum(&output.0)?;
+        builder.append_datum(&output.1)?;
+
+        Ok(())
+    }
+
+    #[inline(always)]
+    fn rollback(builder: &mut ArrayBuilderImpl) {
+        builder.pop().unwrap();
+        builder.pop().unwrap();
+    }
+
+    #[inline(always)]
+    fn finish(writer: SourceStreamChunkRowWriter<'_>) {
+        writer.op_builder.push(Op::UpdateDelete);
+        writer.op_builder.push(Op::UpdateInsert);
+    }
+}
+
 impl SourceStreamChunkRowWriter<'_> {
-    /// Write an `Insert` record to the [`StreamChunk`].
-    ///
-    /// # Arguments
-    ///
-    /// * `self`: Ownership is consumed so only one record can be written.
-    /// * `f`: A closure that produced one [`Datum`] by corresponding [`SourceColumnDesc`].
-    pub fn insert(
+    fn do_action<A: OpAction>(
         self,
-        mut f: impl FnMut(&SourceColumnDesc) -> Result<Datum>,
+        mut f: impl FnMut(&SourceColumnDesc) -> Result<A::Output>,
     ) -> Result<WriteGuard> {
+        // The closure `f` may fail so that a part of builders were appended incompletely.
+        // Loop invariant: `builders[0..appended_idx)` has been appended on every iter ended or loop
+        // exited.
+        let mut appended_idx = 0;
+
         self.descs
             .iter()
             .zip_eq(self.builders.iter_mut())
-            .try_for_each(|(desc, builder)| -> Result<()> {
-                let datum = if desc.skip_parse { None } else { f(desc)? };
-                builder.append_datum(&datum)?;
+            .enumerate()
+            .try_for_each(|(idx, (desc, builder))| -> Result<()> {
+                let output = if desc.skip_parse {
+                    A::DEFAULT_OUTPUT
+                } else {
+                    f(&desc)?
+                };
+                A::apply(builder, output)?;
```

Comment:
Is it possible that `apply` fails and leaves a dirty state in the array builder? 🤣 Rollbacking cannot help in this case.

Reply:
Now apply will not failed.